ARCHITECTURE.md

```diff
@@ -0,0 +1,127 @@
+# Architecture
+
+This document describes high-level architecture of rust-analyzer.
+If you want to familiarize yourself with the code base, you are just
+in the right place!
+
+
+## Code generation
+
+Some of the components of this repository are generated through automatic
+processes. These are outlined below:
+
+- `gen-kinds`: The kinds of tokens are reused in several places, so a generator
+  is used. This process uses [tera] to generate, using data in [grammar.ron],
+  the files:
+  - [ast/generated.rs][ast generated] in `ra_syntax` based on
+    [ast/generated.tera.rs][ast source]
+  - [syntax_kinds/generated.rs][syntax_kinds generated] in `ra_syntax` based on
+    [syntax_kinds/generated.tera.rs][syntax_kinds source]
+
+[tera]: https://tera.netlify.com/
+[grammar.ron]: ./crates/ra_syntax/src/grammar.ron
+[ast generated]: ./crates/ra_syntax/src/ast/generated.rs
+[ast source]: ./crates/ra_syntax/src/ast/generated.tera.rs
+[syntax_kinds generated]: ./crates/ra_syntax/src/syntax_kinds/generated.rs
+[syntax_kinds source]: ./crates/ra_syntax/src/syntax_kinds/generated.tera.rs
+
+
+## Code Walk-Through
+
+### `crates/ra_syntax`
+
+Rust syntax tree structure and parser. See
+[RFC](https://github.com/rust-lang/rfcs/pull/2256) for some design
+notes.
+
+- [rowan](https://github.com/rust-analyzer/rowan) library is used for constructing syntax trees.
+- `grammar` module is the actual parser. It is a hand-written recursive descent parsers, which
+  produced a sequence of events like "start node X", "finish not Y". It works similarly to  [kotlin parser](https://github.com/JetBrains/kotlin/blob/4d951de616b20feca92f3e9cc9679b2de9e65195/compiler/frontend/src/org/jetbrains/kotlin/parsing/KotlinParsing.java),
+  which is a good source for inspiration for dealing with syntax errors and incomplete input. Original [libsyntax parser](https://github.com/rust-lang/rust/blob/6b99adeb11313197f409b4f7c4083c2ceca8a4fe/src/libsyntax/parse/parser.rs)
+  is what we use for the definition of the Rust language.
+- `parser_api/parser_impl` bridges the tree-agnostic parser from `grammar` with `rowan` trees.
+  This is the thing that turns a flat list of events into a tree (see `EventProcessor`)
+- `ast` a type safe API on top of the raw `rowan` tree.
+- `grammar.ron` RON description of the grammar, which is used to
+  generate `syntax_kinds` and `ast` modules, using `cargo gen-kinds` command.
+- `algo`: generic tree algorithms, including `walk` for O(1) stack
+  space tree traversal (this is cool) and `visit` for type-driven
+  visiting the nodes (this is double plus cool, if you understand how
+  `Visitor` works, you understand rust-analyzer).
```

Comment:
It would be nice to link the names to their respective directories/files (e.g. use [ast] instead of `ast` and add a link below)

Reply:
It would be more useful, yeah, but harder to maintain if we move stuff around. Names tend to be relatively more stable, and it's usually not that hard to find a module using GitHub's search/symbol index.